---
title: "Variant_Overlap_2"
author: "Christis Nicolaides"
date: "2025-06-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

```{r}
if("librarian" %in% installed.packages()==FALSE) install.packages("librarian")
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install(version = "3.20")
librarian::shelf(tidyverse, purrr, BiocManager, AnnotationDbi, 
                 qvalue, GO.db, org.Hs.eg.db, reshape2, UniProt.ws, biomaRt, GenomicRanges, rtracklayer)
```

```{r}
#Load data
cad_bed <- read.table(".\\data\\raw\\hglft_genome_31753d_markername.bed", header = FALSE, stringsAsFactors = FALSE)

colnames(cad_bed) <- c("chr", "start", "end", "markername")

pqtl <- as.data.frame(read.csv(".\\data\\raw\\pQTL_discovery.csv"))
pqtl$CHROM <- paste0("chr", pqtl$CHROM)

```

create genomic ranges data
```{r}
gr_cad <- GRanges(
    seqnames = cad_bed$chr,
    ranges   = IRanges(start = cad_bed$start,
                       end   = cad_bed$end),
    # We store the CAD marker name into mcols(gr_cad)$CAD_marker
    CAD_marker = cad_bed$markername
)

print(gr_cad)

pqtl$position <- pqtl$GENPOS..hg38.
pqtl$markername <- pqtl$Variant.ID..CHROM.GENPOS..hg37..A0.A1.imp.v1.

gr_pqtl <-  GRanges(
    seqnames = pqtl$CHROM,
    ranges   = IRanges(start = pqtl$position,
                       end   = pqtl$position),
    # Save the pQTL marker name into the metadata column “pQTL_marker”
    pQTL_marker = pqtl[["markername"]]
)

```

find overlaps
500k
```{r}
hits_500k <- findOverlaps(gr_cad, gr_pqtl,
                           maxgap = 500000)

#create dataframe

overlap_500k <- as_tibble(data.frame(
    CAD_index  = queryHits(hits_500k),
    pQTL_index = subjectHits(hits_500k)
)) %>%
  mutate(
    CAD_marker = mcols(gr_cad)$CAD_marker[CAD_index],
    pQTL_marker = mcols(gr_pqtl)$pQTL_marker[pQTL_index]
  ) %>%
  dplyr::select(CAD_marker, pQTL_marker)

#uplaod to onedrive
saveRDS(
  overlap_500k,
  file = "./data\\processed\\overlap_500k.RDS"
)

write.csv(
  overlap_500k,
  file = "./data\\processed\\overlap_500k.csv",
  row.names = FALSE
)

```

5k
```{r}
hits_5k <- findOverlaps(gr_cad, gr_pqtl, maxgap = 5000)

#dataframe

overlap_5k <- as_tibble(data.frame(
    CAD_index  = queryHits(hits_5k),
    pQTL_index = subjectHits(hits_5k)
)) %>%
  mutate(
    CAD_marker = mcols(gr_cad)$CAD_marker[CAD_index],
    pQTL_marker = mcols(gr_pqtl)$pQTL_marker[pQTL_index]
  ) %>%
  dplyr::select(CAD_marker, pQTL_marker)

#uplaod to onedrive

saveRDS(
  overlap_5k,
  file = "./data\\processed\\overlap_5k.RDS"
)

write.csv(
  overlap_5k,
  file = "./data\\processed\\overlap_5k.csv",
  row.names = FALSE
)
```

1k 
```{r}
hits_1k <- findOverlaps(gr_cad, gr_pqtl, maxgap = 1000)

#create dataframe
overlap_1k <- as_tibble(data.frame(
    CAD_index  = queryHits(hits_1k),
    pQTL_index = subjectHits(hits_1k)
)) %>%
  mutate(
    CAD_marker = mcols(gr_cad)$CAD_marker[CAD_index],
    pQTL_marker = mcols(gr_pqtl)$pQTL_marker[pQTL_index]
  ) %>%
  dplyr::select(CAD_marker, pQTL_marker)

#upload to onedrive
saveRDS(
  overlap_1k,
  file = "./data\\processed\\overlap_1k.RDS"
)

write.csv(
  overlap_1k,
  file = "./data\\processed\\overlap_1k.csv",
  row.names = FALSE
)
```

```{r}
#1bp
hits_3 <- findOverlaps(gr_cad, gr_pqtl, maxgap = 3)

#create dataframe
overlap_3 <- as_tibble(data.frame(
    CAD_index  = queryHits(hits_3),
    pQTL_index = subjectHits(hits_3)
)) %>%
  mutate(
    CAD_marker = mcols(gr_cad)$CAD_marker[CAD_index],
    pQTL_marker = mcols(gr_pqtl)$pQTL_marker[pQTL_index]
  ) %>%
  dplyr::select(CAD_marker, pQTL_marker)

#117 to do coloc on
saveRDS(
  overlap_3,
  file = "./data\\processed\\overlap_3.RDS"
)

write.csv(
  overlap_3,
  file = "./data\\processed\\overlap_3.csv",
  row.names = FALSE
)
```


```{r}
#100kpb look trends in

#map to glycosylation database
#how many pqtld on proteins that have known glycosylaiton sites
#how amny trans genes are enzymes
hits_100k <- findOverlaps(gr_cad, gr_pqtl, maxgap = 100000)

#create dataframe
overlap_100k <- as_tibble(data.frame(
    CAD_index  = queryHits(hits_100k),
    pQTL_index = subjectHits(hits_100k)
)) %>%
  mutate(
    CAD_marker = mcols(gr_cad)$CAD_marker[CAD_index],
    pQTL_marker = mcols(gr_pqtl)$pQTL_marker[pQTL_index]
  ) %>%
  dplyr::select(CAD_marker, pQTL_marker)

saveRDS(
  overlap_100k,
  file = "./data\\processed\\overlap_100k.RDS"
)

write.csv(
  overlap_100k,
  file = "./data\\processed\\overlap_100k.csv",
  row.names = FALSE
)
```

map your variants to the nearest gene located (Bioinformatics.annotated gene) and affected protein and the gene fo that proteib (Assay.Target and Target.UniProt) 

```{r}
pqtl_annotation <- pqtl %>%
  dplyr::select(markername,
         Assay.Target,
         Target.UniProt,
         Bioinfomatic.annotated.gene)%>%
  dplyr::rename(pQTL_marker = markername)

#try 1bp 

annotated_overlap_500k <- merge(overlap_500k,
                                         pqtl_annotation,
                                         by = "pQTL_marker",
                                         all.x = TRUE)

annotated_overlap_5k <- merge(overlap_5k,
                                         pqtl_annotation,
                                         by = "pQTL_marker",
                                         all.x = TRUE)

annotated_overlap_1k <- merge(overlap_1k,
                                         pqtl_annotation,
                                         by = "pQTL_marker",
                                         all.x = TRUE)

annotated_overlap_100k <- merge(overlap_100k,
                                         pqtl_annotation,
                                         by = "pQTL_marker",
                                         all.x = TRUE)

annotated_overlap_3 <- merge(overlap_3,
                                         pqtl_annotation,
                                         by = "pQTL_marker",
                                         all.x = TRUE)

annotated_overlap_100k <- annotated_overlap_100k %>%
  mutate(pQTL_marker_orig = pQTL_marker) %>%
  separate(pQTL_marker, into = c("chr", "pos", "RED", "ALT", "rest"), sep = ":", fill = "right")

annotated_overlap_100k$chr <- ifelse(grepl("^chr", annotated_overlap_100k$chr), annotated_overlap_100k$chr, paste0("chr", annotated_overlap_100k$chr))
annotated_overlap_100k$pos <- as.numeric(annotated_overlap_100k$pos)

gr_100 <- GRanges(seqnames = annotated_overlap_100k$chr, ranges = IRanges(start = annotated_overlap_100k$pos, end = annotated_overlap_100k$pos))
  
annotated_overlap_500k <- annotated_overlap_500k %>%
  mutate(pQTL_marker_orig = pQTL_marker) %>%
  separate(pQTL_marker, into = c("chr", "pos", "RED", "ALT", "rest"), sep = ":", fill = "right")

annotated_overlap_500k$chr <- ifelse(grepl("^chr", annotated_overlap_500k$chr), annotated_overlap_500k$chr, paste0("chr", annotated_overlap_500k$chr))
annotated_overlap_500k$pos <- as.numeric(annotated_overlap_500k$pos)

gr_500 <- GRanges(seqnames = annotated_overlap_500k$chr, ranges = IRanges(start = annotated_overlap_500k$pos, end = annotated_overlap_500k$pos))

chain <- import.chain("./hg38ToHg19.over.chain/hg38ToHg19.over.chain")

lifted_100 <- liftOver(gr_100, chain)%>%
  unlist()
  
annotated_overlap_100k$chr_hg19 <- as.character(seqnames(lifted_100))
annotated_overlap_100k$pos_hg19 <- start(lifted_100)

lifted_500 <- liftOver(gr_500, chain)

annotated_overlap_500k$chr_hg19 <- as.character(seqnames(lifted_500))
annotated_overlap_500k$pos_hg19 <- start(lifted_500)

#print annotate_3 and 100 and 500
saveRDS(
  annotated_overlap_3,
  file = "./data\\processed\\annotated_overlap_3.RDS"
)

write.csv(
  annotated_overlap_3,
  file = "./data\\processed\\annotated_overlap_3.csv",
  row.names = FALSE
)

saveRDS(
  annotated_overlap_100k,
  file = "./data\\processed\\annotated_overlap_100k.RDS"
)

write.csv(
  annotated_overlap_100k,
  file = "./data\\processed\\annotated_overlap_100k.csv",
  row.names = FALSE
)

saveRDS(
  annotated_overlap_500k,
  file = "./data\\processed\\annotated_overlap_500k.RDS"
)

write.csv(
  annotated_overlap_500k,
  file = "./data\\processed\\annotated_overlap_500k.csv",
  row.names = FALSE
)

```

Upload the enzyme and lectin data
```{r}
data_filt_glyenz <- read.csv("./data\\processed\\data_filt_glyenz.csv")
data_gly <-read.csv("./data\\processed\\data_gly_final.csv")

#check overlap on annotated gene with enzyme data

overlap_gene_500k <- unique(annotated_overlap_500k$Bioinfomatic.annotated.gene)

overlap_gene_100k <- unique(annotated_overlap_100k$Bioinfomatic.annotated.gene)

overlap_gene_5k <- unique(annotated_overlap_5k$Bioinfomatic.annotated.gene)

overlap_gene_1k <- unique(annotated_overlap_1k$Bioinfomatic.annotated.gene)

overlap_gene_3 <- unique(annotated_overlap_3$Bioinfomatic.annotated.gene)

common_overlap_enz_500k <- intersect(overlap_gene_500k, data_filt_glyenz$enz_gene)

common_overlap_enz_100k <- intersect(overlap_gene_100k, data_filt_glyenz$enz_gene)

common_overlap_enz_5k <- intersect(overlap_gene_5k, data_filt_glyenz$enz_gene)

common_overlap_enz_1k <- intersect(overlap_gene_1k, data_filt_glyenz$enz_gene)

common_overlap_enz_3 <- intersect(overlap_gene_3, data_filt_glyenz$enz_gene)

print(overlap_gene_3)

print(common_overlap_enz_100k)

#check overlap of assay target with targets in enzyme 

overlap_target_500k <- unique(annotated_overlap_500k$Assay.Target)

overlap_target_100k <- unique(annotated_overlap_100k$Assay.Target)

overlap_target_5k <- unique(annotated_overlap_5k$Assay.Target)

overlap_target_1k <- unique(annotated_overlap_1k$Assay.Target)

overlap_target_3 <- unique(annotated_overlap_3$Assay.Target)

common_overlap_target_500k <- intersect(overlap_target_500k, data_filt_glyenz$Gene_name)

common_overlap_target_100k <- intersect(overlap_target_100k, data_filt_glyenz$Gene_name)

common_overlap_target_5k <- intersect(overlap_target_5k, data_filt_glyenz$Gene_name)

common_overlap_target_1k <- intersect(overlap_target_1k, data_filt_glyenz$Gene_name)

common_overlap_target_3 <- intersect(overlap_target_3, data_filt_glyenz$Gene_name)

#overlap with glycosylation in general?


print(common_overlap_target_100k)

common_targets_list <- list()

genes <- c("ST3GAL4", "GALNT4", "B3GNT3", "B3GNT8", "FUT9", "ST3GAL5", "B4GALNT2")

 for (gene in genes) {
  gene_targets <- data_filt_glyenz %>%
    filter(enz_gene == gene) %>%
    distinct(Gene_name)
  
  common_targets <- intersect(gene_targets$Gene_name, common_overlap_target_100k)
  
  # Store in list, named by gene
  common_targets_list[[gene]] <- common_targets
}

# To view the results:
print(common_targets_list)

ST3GAL4_targets <- data_filt_glyenz %>%
  filter(enz_gene == "ST3GAL4") %>%
  distinct(Gene_name)

ST3GAL4_common_targets_100k <- intersect(ST3GAL4_targets$Gene_name, common_overlap_target_100k)
print(ST3GAL4_common_targets_100k)
```

